File: 0208/push.c
#include <stdio.h>
#include <wiringPi.h>

#define SW1 3
#define SW2 4
#define SW3 5
#define BUZZER_PIN 1

const int melody[] = {0, 262, 294, 330, 349, 392, 440, 494, 523, 587};

int airplane[] = {3, 2, 1, 2, 3, 3, 3, 2, 2, 2, 3, 3, 3, \
                  3, 2, 1, 2, 3, 3, 3, 2, 2, 3, 2, 1};

int cornhamonica[] = {1, 3, 2, 1, 5, 6, 5, 5, 6, 8, 7, 6, 5, 3, 4, 5, 6, \
                  5, 5, 3, 1, 2, 2, 2, 3, 2, 1, 3, 2, 1, 5, 6, 5, 5, 6, 8, 7, 6, 5, \
                  8, 9, 8, 6, 5, 6, 5, 3, 2, 2, 5, 5, 3, 1, 2, 3, 4, 5, 6, 7, 8, \
                  9, 1, 7, 6, 5, 1, 3, 5, 8, 8, 5, 3, 1, 2, 2, 5, 5, 1};

int foxrain[] = {3,2,3,2,3,5,3,3,2,3,2,3,5,3,3,2,3,4,5,3,2,3,4,5,3,2,3,4,5,7,5};
void playMelody(int length);
void playMelody2(int length);
void playMelody3(int length);
void stop_song();

int main(void){
    if (wiringPiSetup() == -1)
        return 1;
    
    pinMode(SW1, INPUT);
    pinMode(SW2, INPUT);
    pinMode(SW3, INPUT);
    pinMode(BUZZER_PIN, PWM_OUTPUT);
    pwmSetClock(19);
    pwmSetMode(PWM_MODE_MS);

    while (1) {
        // 버튼이 눌리면 노래를 재생
        if (digitalRead(SW2) == LOW) {
            printf("노래 1 재생");
            playMelody(sizeof(airplane) / sizeof(airplane[0]));
            // 버튼이 떼질 때까지 대기
            //while (digitalRead(SW2) == LOW);
        }
         if (digitalRead(SW1) == LOW) {
            printf("노래 1 재생");
            playMelody2(sizeof(cornhamonica) / sizeof(cornhamonica[0]));
            // 버튼이 떼질 때까지 대기
            //while (digitalRead(SW1) == LOW);
         }
         if (digitalRead(SW3) == LOW) {
            printf("노래 1 재생");
            playMelody3(sizeof(foxrain) / sizeof(foxrain[0]));
            // 버튼이 떼질 때까지 대기
            //while (digitalRead(SW3) == LOW);
         }
    }

    return 0;
}
void playMelody(int length) {
    int i;
    for(i=0;i<length;i++) {
        if(digitalRead(SW1) == 0) {
			stop_song();
			playMelody2(sizeof(cornhamonica) / sizeof(cornhamonica[0]));
			break;
		}
		else if(digitalRead(SW3) == 0) {
			stop_song();
			playMelody3(sizeof(foxrain) / sizeof(foxrain[0]));
			break;
		}
        pwmSetRange(1000000/melody[airplane[i]]);   
        pwmWrite(BUZZER_PIN, 1000000/melody[airplane[i]]/20);
        if(i==6 || i==9 || i==12 || i==19 || i==24)delay(1000);
        else delay(500); // 각 음을 0.5초 동안 재생
    }
    pwmWrite(BUZZER_PIN, 0);
}

void playMelody2(int length) {
    int i;
    for(i=0;i<length;i++) {
        if(digitalRead(SW2) == 0) {
			stop_song();
			playMelody(sizeof(airplane) / sizeof(airplane[0]));
			break;
		}
		else if(digitalRead(SW3) == 0) {
			stop_song();
			playMelody3(sizeof(foxrain) / sizeof(foxrain[0]));
			break;
		}
        pwmSetRange(1000000/melody[cornhamonica[i]]);
        pwmWrite(BUZZER_PIN, 1000000/melody[cornhamonica[i]]/20);
        if(i==13 || i==26 || i==39 || i==52 || i==65)delay(1000);
        else delay(500); // 각 음을 0.5초 동안 재생
    }
    pwmWrite(BUZZER_PIN, 0);
}

void playMelody3(int length) {
    int i;
    for(i=0;i<length;i++) {
        if(digitalRead(SW1) == 0) {
			stop_song();
			playMelody2(sizeof(cornhamonica) / sizeof(cornhamonica[0]));
			break;
		}
		else if(digitalRead(SW2) == 0) {
			stop_song();
			playMelody(sizeof(airplane) / sizeof(airplane[0]));
			break;
		}
        pwmSetRange(1000000/melody[foxrain[i]]);
        pwmWrite(BUZZER_PIN, 1000000/melody[foxrain[i]]/20);
        if(i==3 || i==10) delay(400);
		else if(i==6 || i==13 || i==18 || i==23) delay(800);
        else delay(500); // 각 음을 0.5초 동안 재생
    }
    pwmWrite(BUZZER_PIN, 0);
}

void stop_song(){
    pwmWrite(BUZZER_PIN, 0);
}

File: 0208/push.py
import RPi.GPIO as GPIO
import time

SW1 = 22
SW2 = 23
SW3 = 24
BUZZER_PIN = 18  # 부저에 연결된 GPIO 핀 번호
GPIO.setmode(GPIO.BCM)
GPIO.setup(BUZZER_PIN, GPIO.OUT)
GPIO.setup(SW1, GPIO.IN)
GPIO.setup(SW2, GPIO.IN)
GPIO.setup(SW3, GPIO.IN)

pwm = GPIO.PWM(BUZZER_PIN, 1000)  # PWM 주파수 설정 (1kHz)

melody = [0, 262, 294, 330, 349, 392, 440, 494, 523, 587]

airplane_note = [3, 2, 1, 2, 3, 3, 3, 2, 2, 2, 3, 3, 3, \
            3, 2, 1, 2, 3, 3, 3, 2, 2, 3, 2, 1]

cornhamonica_note = [1, 3, 2, 1, 5, 6, 5, 5, 6, 8, 7, 6, 5, 3, 4, 5, 6, \
                5, 5, 3, 1, 2, 2, 2, 3, 2, 1, 3, 2, 1, 5, 6, 5, 5, 6, 8, 7, 6, 5, \
                8, 9, 8, 6, 5, 6, 5, 3, 2, 2, 5, 5, 3, 1, 2, 3, 4, 5, 6, 7, 8, \
                9, 1, 7, 6, 5, 1, 3, 5, 8, 8, 5, 3, 1, 2, 2, 5, 5, 1]

foxrain_note = [3, 2, 3, 2, 3, 5, 3, 3, 2, 3, 2, 3, 5, 3, 3, 2, 3, 4, 5, 3, 2, 3, 4, 5, 3, 2, 3, 4, 5, 7, 5]


def airplane(melody_notes):
    pwm.start(10)  # 듀티 사이클을 50%로 설정하여 소리 재생
    for note in melody_notes:
        if GPIO.input(SW1) == GPIO.LOW:
            return 
        if GPIO.input(SW3) == GPIO.LOW:
            return# 다른 버튼이 눌렸을 경우 노래 중지
        frequency = melody[note]
        pwm.ChangeFrequency(frequency)
        if note in [6, 9, 12, 19, 24]:
            time.sleep(1.0)
        else:
            time.sleep(0.5)
    pwm.stop()

def cornhamonica(melody_notes):
    pwm.start(10)  # 듀티 사이클을 50%로 설정하여 소리 재생

    for note in melody_notes:
        if GPIO.input(SW2) == GPIO.LOW:
            return 
        if GPIO.input(SW3) == GPIO.LOW:
            return
        frequency = melody[note]
        pwm.ChangeFrequency(frequency)
        if note in [13, 26, 39, 52, 65]:
            time.sleep(1.0)
        else:
            time.sleep(0.5)
    pwm.stop()

def foxrain(melody_notes):
    pwm.start(10)  # 듀티 사이클을 50%로 설정하여 소리 재생
    for note in melody_notes:
        if GPIO.input(SW1) == GPIO.LOW:
            return 
        if GPIO.input(SW2) == GPIO.LOW:
            return
        frequency = melody[note]
        pwm.ChangeFrequency(frequency)
        if note in [6, 13, 18, 23]:
            time.sleep(1.0)
        else:
            time.sleep(0.5)
    pwm.stop()


def main():
    while True:
        if GPIO.input(SW2) == GPIO.LOW:
            print("노래 1 재생")
            airplane(airplane_note)
            time.sleep(0.1)  # 디바운스
        if GPIO.input(SW1) == GPIO.LOW:
            print("노래 2 재생")
            cornhamonica(cornhamonica_note)
            time.sleep(0.1)  # 디바운스
        if GPIO.input(SW3) == GPIO.LOW:
            print("노래 3 재생")
            foxrain(foxrain_note)
            time.sleep(0.1)  # 디바운스



if __name__ == "__main__":
    main()
